Objects/typeobject.c

```diff
@@ -8831,6 +8831,71 @@ super_getattro(PyObject *self, PyObject *name)
     return PyObject_GenericGetAttr(self, name);
 }
 
+static int
+super_setattro(PyObject *self, PyObject *name, PyObject *value)
+{
+    superobject *su = (superobject *)self;
+    int skip = (su->obj_type == NULL);
+
+    if (!skip) {
+        /* We don't allow overwriting __class__. */
+        skip = (PyUnicode_Check(name) &&
+            PyUnicode_GET_LENGTH(name) == 9 &&
+            PyUnicode_CompareWithASCIIString(name, "__class__") == 0);
+    }
+
+    if (!skip) {
+        PyObject *mro, *descr, *tmp, *dict;
+        PyTypeObject *starttype;
+        descrsetfunc f;
+        Py_ssize_t i, n;
+        int done = 0, result = 0;
+
+        starttype = su->obj_type;
+        mro = starttype->tp_mro;
+
+        if (mro == NULL)
+            n = 0;
+        else {
+            assert(PyTuple_Check(mro));
+            n = PyTuple_GET_SIZE(mro);
+        }
+        for (i = 0; i < n; i++) {
+            if ((PyObject *)(su->type) == PyTuple_GET_ITEM(mro, i))
+                break;
+        }
+        i++;
+        descr = NULL;
+        /* keep a strong reference to mro because starttype->tp_mro can be
+           replaced during PyDict_GetItem(dict, name)  */
+        Py_INCREF(mro);
+        for (; i < n; i++) {
+            tmp = PyTuple_GET_ITEM(mro, i);
+            if (PyType_Check(tmp))
+                dict = ((PyTypeObject *)tmp)->tp_dict;
+            else
+                continue;
+            descr = PyDict_GetItem(dict, name);
+            if (descr != NULL) {
+                Py_INCREF(descr);
+                f = Py_TYPE(descr)->tp_descr_set;
+                if ((f != NULL) && PyDescr_IsData(descr)) {
+                    /* We found a data descriptor: */
```

Comment:
Seems like the behavior if there isn't a data descriptor is that we fall back to `PyObject_GenericSetAttr`, which will fail because `super()` doesn't allow setting any attributes. Shouldn't we instead set the attr on the object super is proxying for?

Use case to think about: We're using `super().x = ...` to set a `@property` on a base class. Then the base class is refactored so that `x` is a regular attribute, not a `@property`. As a user, I would expect `super().x = ...` to continue to work as before.